launcher/MMCTime.cpp

```diff
@@ -78,25 +78,124 @@ QString Time::humanReadableDuration(double duration, int precision) {
     if (hc) {
         if (dc)
             os << " ";
-        os << qSetFieldWidth(2) << hc << QObject::tr("h"); // hours
+        os << qSetFieldWidth(2) << hc << QObject::tr("h");  // hours
     }
     if (mc) {
         if (dc || hc)
             os << " ";
-        os << qSetFieldWidth(2) << mc << QObject::tr("m"); // minutes
+        os << qSetFieldWidth(2) << mc << QObject::tr("m");  // minutes
     }
     if (dc || hc || mc || sc) {
         if (dc || hc || mc)
             os << " ";
-        os << qSetFieldWidth(2) << sc << QObject::tr("s"); // seconds
+        os << qSetFieldWidth(2) << sc << QObject::tr("s");  // seconds
     }
     if ((msc && (precision > 0)) || !(dc || hc || mc || sc)) {
         if (dc || hc || mc || sc)
             os << " ";
-        os << qSetFieldWidth(0) << qSetRealNumberPrecision(precision) << msc << QObject::tr("ms"); // miliseconds
+        os << qSetFieldWidth(0) << qSetRealNumberPrecision(precision) << msc << QObject::tr("ms");  // miliseconds
     }
 
     os.flush();
 
     return outStr;
+}
+
+QString Time::humanReadableDuration(int64_t duration, const QString& fmt)
+{
+    using days = std::chrono::duration<double, std::ratio<86400>>;
+    using daysR = std::chrono::duration<int, std::ratio<86400>>;
+    using hours = std::chrono::duration<double, std::chrono::hours::period>;
+    using minutes = std::chrono::duration<double, std::chrono::minutes::period>;
+    auto std_duration = std::chrono::duration<double>(double(duration));
+
+    QString outStr;
+    QTextStream os(&outStr);
+
+    QChar option = QChar::Null;
+    bool inOption;
+    QString precSeg;
+    QTextStream s(&precSeg);
+
+    QString segment;
+    QTextStream seg(&segment);
+
+    for (auto i = 0; i < fmt.size(); i++) {
+        auto c = fmt[i];
+        if (!inOption) {
+            if (inOption = c == '%'; inOption) {
+                seg.flush();
+                os << segment;
+                seg.reset();
+                segment = "";
+            } else {
+                seg << c;
+            }
+            continue;
+        }
+        if (c == '%') {
+            inOption = false;
+            int precision = 0;
+            s.flush();
+            if (precSeg.startsWith('.')) {
+                precision = precSeg.remove(0, 1).toInt();
+            }
+            s.reset();
+            precSeg = "";
+            double dc = 0;
+            if (option == 'd') {
```

Comment:
hate to be that person but...couldn't this be a `switch` statement? i think it'd just look a little nicer :p

Reply:
changed most of the code there to use switch